src/fun/transform/resugar_string.rs

```diff
@@ -7,101 +7,132 @@ impl Term {
+  }
+
+  /// Attempts to resugar a common cons term.
+  fn try_resugar_cons_common(&mut self, build_string_fn: fn(&Term, String) -> Option<String>) -> bool {
+    let Term::App { tag: Tag::Static, fun, arg: tail } = self else { return false };
+    let Term::App { tag: Tag::Static, fun: inner_fun, arg: head } = fun.as_mut() else { return false };
+    let (Term::Ref { nam }, Term::Num { val: Num::U24(head_val) }) = (inner_fun.as_ref(), head.as_ref())
+    else {
+      return false;
+    };
+
+    if nam != builtins::SCONS {
+      return false;
+    };
+
+    let head_char = char::from_u32(*head_val).unwrap_or(char::REPLACEMENT_CHARACTER);
+    build_string_fn(tail, head_char.to_string()).map(|str| *self = Term::str(&str)).is_some()
+  }
+
+  /// common string building function that uses custom and default cons handlers.
+  /// pattern keep if tree mimics the shape of the AST visualize the structure #713
+  fn build_string<F>(term: &Term, mut s: String, cons_handler: F) -> Option<String>
+  where
+    F: Fn(&Term) -> Option<(char, &Term)>,
+  {
     maybe_grow(|| {
-      // Search for a String/cons pattern in the term and try to build a string from that point on.
-      // If successful, replace the term with the string.
-      // If not, keep as-is.
-
-      // Nil: String/nil
-      if let Term::Ref { nam } = self {
-        if nam == builtins::SNIL {
-          *self = Term::str("");
+      let mut current = term;
+      loop {
+        // Nil: String/nil
+        if let Term::Ref { nam } = current {
+          if nam == builtins::SNIL {
+            return Some(s);
+          }
         }
-      }
-      // Cons: @* @c (c <num> <str>)
-      if let Term::Lam { tag: Tag::Static, pat, bod } = self {
-        if let Pattern::Var(None) = pat.as_mut() {
-          if let Term::Lam { tag: Tag::Static, pat, bod } = bod.as_mut() {
-            if let Pattern::Var(Some(var_lam)) = pat.as_mut() {
-              if let Term::App { tag: Tag::Static, fun, arg: tail } = bod.as_mut() {
-                if let Term::App { tag: Tag::Static, fun, arg: head } = fun.as_mut() {
-                  if let Term::Var { nam: var_app } = fun.as_mut() {
-                    if let Term::Num { val: Num::U24(head) } = head.as_mut() {
-                      if var_lam == var_app {
-                        let head = char::from_u32(*head).unwrap_or(char::REPLACEMENT_CHARACTER);
-                        if let Some(str) = build_string_scott(tail, head.to_string()) {
-                          *self = Term::str(&str);
-                        } else {
-                          // Not a string term, keep as-is.
+
+        // Try custom cons handler
+        if let Some((head, tail)) = cons_handler(current) {
+          s.push(head);
+          current = tail;
+          continue;
+        }
+
+        // Cons: @x (x CONS_TAG <num> <str>)
+        if let Term::Lam { tag: Tag::Static, pat, bod } = current {
```

Comment:
This is specifically the "num-scott" encoding, you shouldn't apply it for all cases